### DISCUSSION TITLE
Extract the original MIP model within the observation function

### DISCUSSION THREAD

**Discussion by zhejiangzhoulei** [ds4dm/ml4co-competition#32] [Q&A] [answered]
Hi all,
I am trying to extract the original MIP model within the observation function.
It's easy to extract the variables: 
```
        # extract info of var
        variables = m.getVars(transformed=False)
        inf = m.infinity()

        lbs = np.asarray([v.getLbLocal() for v in variables])
        ubs = np.asarray([v.getUbLocal() for v in variables])
        has_lb = np.asarray([lb > -inf for lb in lbs])
        has_ub = np.asarray([ub < inf for ub in ubs])
        v_id = np.asarray([v.getIndex() for v in variables])
        v_name = np.asarray([v.name for v in variables])
        v_type = np.asarray([v.vtype() for v in variables])
```

However, from the PySCIPOpt API, I do not find the method to fetch the original constraints and reconstruct the original MIP problem.

Any help or advice would be greatly appreciated.

Best

**Comment by fsikora** [comment]
You can check the answers provided here : https://github.com/ds4dm/ml4co-competition/discussions/18

**Reply by zhejiangzhoulei** [reply]
@fsikora thank you for your response. #18 gives the methods to fetch the LP model, but I want to obtain the IP model.

**Reply by fsikora** [reply]
see https://github.com/ds4dm/ml4co-competition/discussions/18#discussioncomment-995362, it seems it will come later

**Reply by zhejiangzhoulei** [reply]
@fsikora thanks~ :)

**Comment by gasse** [comment, marked as answer]
Hi @zhejiangzhoulei ,

I suppose you are interested in recovering the original MILP for the primal task ? Otherwise the following might not apply.

First, I want to point out that what you want is most likely the **transformed** MILP, that is, the original MILP plus any preprocessing made by SCIP (some constraints might have been simplified or added, some variables might have been removed, merged or added, etc.). The solutions that your agent must provide in the primal task are solutions in the transformed variable space, so you might as well extract the problem formulation directly in that space as well in you observation function to avoid any headache.

You should be able to recover the transformed MILP of the current node (i.e., the root node in the primal task) from the PySCIPOpt API by using the following API (as @fsikora also pointed out) :
```python
m = model.as_pyscipopt()
cols = [var.getCol() for var in m.getVars(transformed=True) if var.isInLP()]  # to get local LP columns
rows = m.getLPRowsData()  # to get local LP rows

# for each row in the LP (lhs <= wTx + cst <= rhs)
row.getLhs()       # left-hand-side of the row (lhs)
row.getRhs()       # right-hand-side of the row (rhs)
row.getConstant()  # constant of a row (cst)
row.getCols()      # columns with a non-zero entry in the row (i where w_i != 0)
row.getVals()      # the corresponding non-zero entries (w_i where w_i != 0)

# for each column in the LP (x_i)
col.getLPPos()         # position (id) of the column in the LP
col.getLb()            # lower bound of the column (lb <= x_i)
col.getUb()            # upper bound of the column (x_i <= ub)
col.getVar().getObj()  # objective coefficient of the column (c_i)
col.getVar().vtype()   # variable type (BINARY, INTEGER, IMPLINT or CONTINUOUS)
```

Those calls shall allow you to reconstruct the transformed MILP, for example if you want to feed it into another off-the-shelf search method for MILPs.

Note: don't forget to extract the variable indexes as well (`col.getVar().getIndex()`), to make sure you do not loose the mapping between the environment variables (SCIP's variable indexes in the transformed problem), and the variable values that some another software would give you.

Hope that helps. Let us know if you encounter any difficulty.

Best,
Maxime

**Reply by zhejiangzhoulei** [reply]
@gasse thank you so much!!!
